dotcom-rendering/src/model/buildLightboxImages.ts

```diff
@@ -110,78 +125,41 @@ export const buildLightboxImages = (
-		block.elements.forEach((element) => {
-			switch (element._type) {
-				case 'model.dotcomrendering.pageElements.ImageBlockElement': {
-					const lightboxImage = buildLightboxImage(element);
-					if (lightboxImage === undefined) break;
-					if (
-						format.design === 'LiveBlogDesign' ||
-						format.design === 'DeadBlogDesign'
-					) {
-						lightboxImage.blockId = block.id;
-						lightboxImage.firstPublished =
-							block.blockFirstPublished;
-					}
-					lightboxImages.push(lightboxImage);
-					break;
-				}
-				case 'model.dotcomrendering.pageElements.MultiImageBlockElement': {
-					element.images.forEach((multiImage) => {
-						const lightboxImage = buildLightboxImage(multiImage);
-						if (lightboxImage === undefined) return;
-						if (
-							format.design === 'LiveBlogDesign' ||
-							format.design === 'DeadBlogDesign'
-						) {
-							lightboxImage.blockId = block.id;
-							lightboxImage.firstPublished =
-								block.blockFirstPublished;
-						}
-						lightboxImages.push(lightboxImage);
-					});
-					break;
-				}
-				default:
-					break;
-			}
-		});
-	});
+	const lightboxImages = mainMediaElements
+		.flatMap<ImageForLightbox>((element) => {
+			if (isImage(element)) return buildLightboxImage(element) ?? [];
+			return [];
+		})
+		.concat(
+			blocks.flatMap<ImageForLightbox>((block) =>
+				block.elements.flatMap<ImageForLightbox>((element) =>
+					getImages(element).flatMap<ImageForLightbox>(
+						(multiImage) => {
+							const lightboxImage =
+								buildLightboxImage(multiImage);
+							if (lightboxImage === undefined) return [];
+							return isBlog(format.design)
+								? {
+										...lightboxImage,
+										blockId: block.id,
+										firstPublished:
+											block.blockFirstPublished,
+								  }
+								: lightboxImage;
+						},
+					),
+				),
+			),
+		);
+
 	// On gallery articles the main media is often repeated as an element in the article body so
 	// we deduplicate the array here
-	const alreadySeen: string[] = [];
-	let uniqueImages: ImageForLightbox[] = [];
-	lightboxImages.forEach((image) => {
-		const imageId = decideImageId(image);
-		if (!imageId) {
-			// It's unlikely that we would not have an imageId here but if we do we fail
-			// open, passing the image through
-			uniqueImages.push(image);
-			return;
-		}
-		if (alreadySeen.includes(imageId)) {
-			// Replace the element with the duplicated one. We do this because we want to favour
-			// the higher quality body images you get with galleries.
-			uniqueImages = uniqueImages.filter(
-				(thisImage) => decideImageId(thisImage) !== imageId,
-			);
-			uniqueImages.push(image);
-		} else {
-			// This image is not duplicated so we pass it through
-			uniqueImages.push(image);
-			alreadySeen.push(imageId);
-		}
-	});
-	return uniqueImages;
+	return [
+		...new Map<string, ImageForLightbox>(
+			lightboxImages.map(
+				(image) =>
+					[decideImageId(image) ?? randomUUID(), image] as const,
```

Comment:
not really about this PR, but I was wondering what `as const` is doing here? I couldn't see a difference in the type of the return value when I removed it, so I'm just curious!

Reply:
Ah good question I wasn't sure that it could be narrow enough for the Map constructor to figure things out, but if unnecessary I'll remove.

I though that it would narrow from `(string | LightboxImage)[]` to `[string, LightboxImage]` but forgot to actually check!